File: AlgorithmReview/Programmers/Lv3/단어 변환.swift
//
//  단어 변환.swift
//  AlgorithmReview
//
//  Created by 박소윤 on 2023/05/06.
//

import Foundation

func solution(_ begin:String, _ target:String, _ words:[String]) -> Int {
    
    var answer = Int.max
    var visit = [Bool](repeating: false, count: words.count)
    
    func isValid(_ a: String, _ b: String) -> Bool {
        let a = Array(a)
        let b = Array(b)
        var diff = 0
        for i in 0..<a.count {
            if a[i] != b[i] {
                diff += 1
            }
            if diff > 1 {
                return false
            }
        }
        return true
    }
    
    func dfs(_ w: String, _ depth: Int) {
        if target == w {
            answer = min(answer, depth)
            return 
        }
        if depth >= answer {
            return
        }
        for (i, temp) in words.enumerated() {
            if visit[i] { continue }
            if isValid(temp, w) {
                visit[i] = true
                dfs(temp, depth + 1)
                visit[i] = false
            }
        }
    }
    
    dfs(begin, 0)
    
    if answer == Int.max {
        return 0
    }
    return answer
}

File: Python/단어 변환.py
def solution(begin, target, words):
    
    maxSize = 2**63-1
    
    def isValid(a, b):
        
        a = list(a)
        b = list(b)
        
        diff = 0
        for i in range(len(a)):
            if a[i] != b[i]:
                diff += 1
            if diff > 1 :
                return False
        return True
    
    global answer
    answer = maxSize
    
    global visit
    visit = [False] * len(words)
    
    def dfs(w, depth):
        
        if w == target:
            global answer
            answer = min(answer, depth)
            return
        if depth >= answer:
            return
        
        for (i, temp) in enumerate(words):
            global visit
            if visit[i]:
                continue
            if isValid(temp, w):
                visit[i] = True
                dfs(temp, depth+1)
                visit[i] = False
    
    dfs(begin, 0)
    
    if answer == maxSize:
        return 0
    return answer
